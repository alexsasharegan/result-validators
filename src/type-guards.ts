export function is_string(x: any): x is string {
  return typeof x == "string";
}

export function is_not_string<T>(x: T): x is Exclude<T, string> {
  return !is_string(x);
}

export function is_number(x: any): x is number {
  return typeof x == "number";
}

export function is_not_number<T>(x: T): x is Exclude<T, number> {
  return !is_number(x);
}

export function is_object(x: any): x is object {
  return typeof x == "object";
}

export function is_not_object<T>(x: T): x is Exclude<T, object> {
  return !is_object(x);
}

export function is_boolean(x: any): x is boolean {
  return typeof x == "boolean";
}

export function is_not_boolean<T>(x: T): x is Exclude<T, boolean> {
  return !is_boolean(x);
}

export function is_function(x: any): x is Function {
  return typeof x == "function";
}

export function is_not_function<T>(x: T): x is Exclude<T, Function> {
  return !is_function(x);
}

export function is_symbol(x: any): x is Symbol {
  return typeof x == "symbol";
}

export function is_not_symbol<T>(x: T): x is Exclude<T, Symbol> {
  return !is_symbol(x);
}

export function is_undefined(x: any): x is undefined {
  return typeof x == "undefined";
}

export function is_not_undefined<T>(x: T): x is Exclude<T, undefined> {
  return !is_undefined(x);
}
