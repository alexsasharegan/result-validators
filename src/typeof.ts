export function is_string(x: any): x is string {
  return typeof x == "string";
}

export function is_number(x: any): x is number {
  return typeof x == "number";
}

export function is_object(x: any): x is object {
  return typeof x == "object";
}

export function is_boolean(x: any): x is boolean {
  return typeof x == "boolean";
}

export function is_function(x: any): x is Function {
  return typeof x == "function";
}
