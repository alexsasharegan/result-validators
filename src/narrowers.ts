import { Option, Result, Ok, Err } from "safe-types";
import { is_string, is_number, is_boolean } from "./type-guards";

/**
 * A `TypeNarrower` takes any value, and returns a result with the
 * type narrowed & asserted on success, or type error string on failure.
 */
export type TypeNarrower<T> = (x: any) => Result<T, string>;

/**
 * Narrow any type to a `string`.
 */
export const asString: TypeNarrower<string> = x =>
  Option.of(x)
    .narrow(is_string)
    .match({
      None: () => Err(`value is not of type 'string'`),
      Some: Ok,
    });

/**
 * A loose type narrower that **does not** check for `NaN` cases.
 */
export const asAnyNumber: TypeNarrower<number> = x =>
  Option.of(x)
    .narrow(is_number)
    .match({
      None: () => Err(`value is not of type 'number'`),
      Some: Ok,
    });

/**
 * A strict type narrower that fails on `NaN` values.
 */
export const asNumber: TypeNarrower<number> = x =>
  asAnyNumber(x).and_then(x => {
    if (Number.isNaN(x)) {
      return Err(`number is NaN`);
    }

    return Ok(x);
  });

/**
 * Narrow any type to a `boolean`.
 */
export const asBool: TypeNarrower<boolean> = x =>
  Option.of(x)
    .narrow(is_boolean)
    .match({
      None: () => Err(`value is not of type 'boolean'`),
      Some: Ok,
    });

/**
 * Narrow any type to a `Array<any>`.
 */
export const asArray: TypeNarrower<any[]> = x =>
  Option.of(x)
    .narrow(Array.isArray)
    .match({
      None: () => Err(`value is not an Array`),
      Some: Ok,
    });

/**
 * Narrow any type to a `number` that is an integer.
 */
export const asInt: TypeNarrower<number> = x =>
  Option.of(x)
    .narrow(is_number)
    .filter(Number.isInteger)
    .match({
      None: () => Err(`value is not an integer`),
      Some: Ok,
    });

/**
 * Narrow any type to an instance of a custom class/constructor.
 */
export const asInstanceOf: <C extends new (...args: any[]) => any>(
  ctor: C
) => TypeNarrower<InstanceType<C>> = ctor => x =>
  Option.of(x)
    .filter(x => x instanceof ctor)
    .match({
      None: () => Err(`value is not an instance of ${ctor.name}`),
      Some: Ok,
    });
