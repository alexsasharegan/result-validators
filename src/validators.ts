import { Option, Result, Ok, Err } from "safe-types";
import { is_string, is_number } from "./typeof";

export interface Validator<T> {
  (x: T): Result<T, string>;
}

/**
 * Given any type, returns `Ok(string)` or `Err(string)`.
 */
export function asString(s: any): Result<string, string> {
  return Option.of(s)
    .filter(is_string)
    .match({
      None: () => Err(`value is not of type 'string'`),
      Some: Ok,
    });
}

export function asInt(x: any): Result<number, string> {
  return Option.of(x)
    .filter(is_number)
    .filter(Number.isInteger)
    .match({
      None: () => Err(`value is not an integer`),
      Some: Ok,
    });
}

/**
 * Given a string, returns `Ok(string)`
 * if that string is URL safe or `Err(string)`.
 */
export function asURLSafe(s: string): Result<string, string> {
  if (encodeURIComponent(s) !== s) {
    return Err(`text contains invalid characters: '${s}'`);
  }

  return Ok(s);
}

/**
 * Given a string, returns `Ok(string)`
 * if that string is trimmed or `Err(string)`.
 */
export function asTrimmed(s: string): Result<string, string> {
  if (s.trim() !== s) {
    return Err(`text contains untrimmed whitespace: '${s}'`);
  }

  return Ok(s);
}

/**
 * Given a string, returns `Ok(string)`
 * if that string is lowercase or `Err(string)`.
 */
export function asLowerCased(s: string): Result<string, string> {
  if (s.toLowerCase() !== s) {
    return Err(`text is not lowercase: '${s}'`);
  }

  return Ok(s);
}

/**
 * Given a string, returns `Ok(string)`
 * if that string only contains hex chars or `Err(string)`
 */
export function asHex(s: string): Result<string, string> {
  for (let i = 0, len = s.length, code: number; i < len; i++) {
    code = s.charCodeAt(i);
    // 0-9 are code points 48-57
    // A-F are code points 65-70
    // a-f are code points 97-102
    if (
      (code < 48 || code > 57) &&
      (code < 65 || code > 70) &&
      (code < 97 || code > 102)
    ) {
      return Err(`text contains invalid hexadecimal characters: '${s}'`);
    }
  }

  return Ok(s);
}

/**
 * Given a string, returns `Ok(string)`
 * if that string contains only printable chars or `Err(string)`
 */
export function asPrintableChars(s: string): Result<string, string> {
  // Non-printable chars are code points 0-31
  for (let i = 0, len = s.length; i < len; i++) {
    if (s.charCodeAt(i) < 32) {
      return Err(`text contains non-printable characters: '${s}'`);
    }
  }

  return Ok(s);
}

/**
 * Given a string, returns `Ok(string)`
 * if that string only contains lowercase hex chars or `Err(string)`
 */
export function asHexLowerCased(s: string): Result<string, string> {
  return asHex(s).and_then(asLowerCased);
}

/**
 * Given a string, returns `Ok(string)`
 * if that string only contains digits or `Err(string)`
 */
export function asDigit(d: string): Result<string, string> {
  for (let i = 0, len = d.length; i < len; i++) {
    // 0-9 are code points 48-57
    if (d.charCodeAt(i) < 48 || d.charCodeAt(i) > 57) {
      return Err(`text contains non-numeric characters: '${d}'`);
    }
  }

  return Ok(d);
}

export function withMax(max: number): Validator<number> {
  return function asMax(x) {
    if (x > max) {
      return Err(`the value cannot exceed ${max}`);
    }

    return Ok(x);
  };
}

export function withMin(min: number): Validator<number> {
  return function asMin(x) {
    if (x < min) {
      return Err(`the value cannot be less than ${min}`);
    }

    return Ok(x);
  };
}

export function betweenRange(min: number, max: number): Validator<number> {
  return function asRange(x) {
    return Ok(x)
      .and_then(withMin(min))
      .and_then(withMax(max));
  };
}

/**
 * Given a length, returns a function that takes a string
 * and returns `Ok(string)` if that string === length or `Err(string)`.
 */
export function withLength(len: number): Validator<string> {
  return function asLength(s) {
    if (s.length !== len) {
      return Err(`the length must be ${len}: found ${s.length}`);
    }

    return Ok(s);
  };
}

/**
 * Given a min/max length, returns a function that takes a string
 * and returns `Ok(string)` if that string length is between min/max or `Err(string)`.
 */
export function withLengthBetweenRange<T>(
  min: number,
  max: number
): Validator<ArrayLike<T>> {
  return function asLengthBetweenRange(s) {
    if (s.length < min || s.length > max) {
      return Err(`the length must be between ${min}-${max}: found ${s.length}`);
    }

    return Ok(s);
  };
}

/**
 * Given a min length, returns a function that takes a string
 * and returns `Ok(string)` if that string is min length or `Err(string)`.
 */
export function withMinLength<T>(min: number): Validator<ArrayLike<T>> {
  return function asMinLength(s) {
    if (s.length < min) {
      return Err(`the length must be at least ${min}: found ${s.length}`);
    }

    return Ok(s);
  };
}

/**
 * Given a max length, returns a function that takes a string
 * and returns `Ok(string)` if that string is not greater than max or `Err(string)`.
 */
export function withMaxLength<T>(max: number): Validator<ArrayLike<T>> {
  return function asMaxLength(s) {
    if (s.length > max) {
      return Err(`the length cannot exceed ${max}: found ${s.length}`);
    }

    return Ok(s);
  };
}

/**
 * Given a byte length, returns a function that takes a string
 * and returns `Ok(string)` if that string === length or `Err(string)`.
 */
export function withByteLength(len: number): Validator<string> {
  return function asByteLength(s) {
    let byteLength = Buffer.from(s, "utf8").byteLength;
    if (byteLength !== len) {
      return Err(`the byte length must be ${len}: found ${byteLength}`);
    }

    return Ok(s);
  };
}

/**
 * Given a min/byte max length, returns a function that takes a string
 * and returns `Ok(string)` if that string length is between min/max or `Err(string)`.
 */
export function withByteLengthBetweenRange(
  min: number,
  max: number
): Validator<string> {
  return function asByteLengthBetweenRange(s) {
    let byteLength = Buffer.from(s, "utf8").byteLength;
    if (byteLength < min || byteLength > max) {
      return Err(
        `the byte length must be between ${min}-${max}: found ${byteLength}`
      );
    }

    return Ok(s);
  };
}

/**
 * Given a min byte length, returns a function that takes a string
 * and returns `Ok(string)` if that string is min length or `Err(string)`.
 */
export function withMinByteLength(min: number): Validator<string> {
  return function asMinByteLength(s) {
    let byteLength = Buffer.from(s, "utf8").byteLength;
    if (byteLength < min) {
      return Err(
        `the byte length must be at least ${min}: found ${byteLength}`
      );
    }

    return Ok(s);
  };
}

/**
 * Given a max byte length, returns a function that takes a string
 * and returns `Ok(string)` if that string is not greater than max or `Err(string)`.
 */
export function withMaxByteLength(max: number): Validator<string> {
  return function asMaxByteLength(s) {
    let byteLength = Buffer.from(s, "utf8").byteLength;
    if (byteLength > max) {
      return Err(`the byte length cannot exceed ${max}: found ${byteLength}`);
    }

    return Ok(s);
  };
}

/**
 * Given a `Set<T>`, returns a function that takes a value `T`
 * and returns `Ok(T)` if the Set contains T or `Err(string)`.
 */
export function oneOfSet<T>(
  set: Set<T>,
  msg: string = `the given value is not one of the expected set of values`
): Validator<T> {
  return function asOneOfSet(x) {
    if (!set.has(x)) {
      return Err(msg);
    }

    return Ok(x);
  };
}
