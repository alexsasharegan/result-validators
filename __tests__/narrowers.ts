import { Ok, Err } from "safe-types";
import * as N from "../src/narrowers";

describe("Narrowers", async () => {
  it("should for strings", async () => {
    let t = {
      success: ["a string", ""],
      fail: [1, false, {}, [], null, undefined],
    };

    for (let ts of t.success) {
      expect(N.asString(ts)).toEqual(Ok(ts));
    }

    for (let tf of t.fail) {
      let r = N.asString(tf);
      expect(r.is_err()).toBe(true);
      expect(r.unwrap_err()).toMatchSnapshot();
    }
  });

  it("should for booleans", async () => {
    let t = {
      success: [true, false],
      fail: [1, "false", "true", {}, [], null, undefined],
    };

    for (let ts of t.success) {
      expect(N.asBool(ts)).toEqual(Ok(ts));
    }

    for (let tf of t.fail) {
      let r = N.asBool(tf);
      expect(r.is_err()).toBe(true);
      expect(r.unwrap_err()).toMatchSnapshot();
    }
  });

  it("should for numbers (loose)", async () => {
    let t = {
      success: [
        0,
        -0,
        1,
        -1,
        Number.MAX_VALUE,
        Number.MAX_SAFE_INTEGER,
        Number.MIN_VALUE,
        Number.MIN_SAFE_INTEGER,
        NaN,
        Number.NaN,
      ],
      fail: ["0", "1", true, false, {}, [], null, undefined],
    };

    for (let ts of t.success) {
      expect(N.asAnyNumber(ts)).toEqual(Ok(ts));
    }

    for (let tf of t.fail) {
      let r = N.asAnyNumber(tf);
      expect(r.is_err()).toBe(true);
      expect(r.unwrap_err()).toMatchSnapshot();
    }
  });

  it("should for numbers (strict)", async () => {
    let t = {
      success: [
        0,
        -0,
        1,
        -1,
        Number.MAX_VALUE,
        Number.MAX_SAFE_INTEGER,
        Number.MIN_VALUE,
        Number.MIN_SAFE_INTEGER,
      ],
      fail: [NaN, Number.NaN, "0", "1", true, false, {}, [], null, undefined],
    };

    for (let ts of t.success) {
      expect(N.asNumber(ts)).toEqual(Ok(ts));
    }

    for (let tf of t.fail) {
      let r = N.asNumber(tf);
      expect(r.is_err()).toBe(true);
      expect(r.unwrap_err()).toMatchSnapshot();
    }
  });

  it("should for ints (strict)", async () => {
    let t = {
      success: [
        0,
        -0,
        1,
        -1,
        5.0,
        10 / 2,
        Number.MAX_SAFE_INTEGER,
        Number.MIN_SAFE_INTEGER,
        Number.MAX_VALUE,
      ],
      fail: [
        0.1,
        -0.1,
        1 / 2,
        NaN,
        Number.NaN,
        "0",
        "1",
        true,
        false,
        {},
        [],
        null,
        undefined,
        Number.MIN_VALUE,
      ],
    };

    for (let ts of t.success) {
      let r = N.asInt(ts);
      r.map_err(err =>
        console.error(err, ts, `(index[${t.success.indexOf(ts)}])`)
      );
      expect(r).toEqual(Ok(ts));
    }

    for (let tf of t.fail) {
      let r = N.asInt(tf);
      r.map(() =>
        console.error(
          `Value should be error:`,
          tf,
          `(at index [${t.fail.indexOf(tf)}])`
        )
      );
      expect(r.is_err()).toBe(true);
      expect(r.unwrap_err()).toMatchSnapshot();
    }
  });

  it("should for arrays", async () => {
    let t = {
      success: [[], [1, 2, 3, 4, 5], [1, true, "string", {}, []]],
      fail: [
        1,
        "false",
        "true",
        {},
        { 0: "", 1: "", length: 2 },
        null,
        undefined,
      ],
    };

    for (let ts of t.success) {
      expect(N.asArray(ts)).toEqual(Ok(ts));
    }

    for (let tf of t.fail) {
      let r = N.asArray(tf);
      expect(r.is_err()).toBe(true);
      expect(r.unwrap_err()).toMatchSnapshot();
    }
  });

  it("should for custom classes", async () => {
    class A {}
    class B extends A {}
    class C extends B {}
    class D {}

    let t = {
      success: [
        { ctor: A, subject: new A() },
        { ctor: A, subject: new B() },
        { ctor: A, subject: new C() },
        { ctor: B, subject: new C() },
        { ctor: D, subject: new D() },
        { ctor: Object, subject: {} },
        { ctor: Object, subject: new D() },
        { ctor: Function, subject() {} },
        { ctor: Object, subject() {} },
      ],
      fail: [
        { ctor: C, subject: new B() },
        { ctor: B, subject: new A() },
        { ctor: D, subject: new A() },
        { ctor: A, subject: {} },
        { ctor: B, subject: {} },
        { ctor: A, subject() {} },
      ],
    };

    for (let ts of t.success) {
      let r = N.asInstanceOf(ts.ctor)(ts.subject);
      r.map_err(err =>
        console.error(err, ts, `(index[${t.success.indexOf(ts)}])`)
      );
      expect(r).toEqual(Ok(ts.subject));
    }

    for (let tf of t.fail) {
      let r = N.asInstanceOf(tf.ctor)(tf.subject);
      r.map(() =>
        console.error(
          `Value should be error:`,
          tf,
          `(at index [${t.fail.indexOf(tf)}])`
        )
      );
      expect(r.is_err()).toBe(true);
      expect(r.unwrap_err()).toMatchSnapshot();
    }
  });
});
