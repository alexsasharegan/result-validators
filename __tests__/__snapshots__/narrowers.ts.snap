// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Narrowers should for arrays 1`] = `"value is not an Array"`;

exports[`Narrowers should for arrays 2`] = `"value is not an Array"`;

exports[`Narrowers should for arrays 3`] = `"value is not an Array"`;

exports[`Narrowers should for arrays 4`] = `"value is not an Array"`;

exports[`Narrowers should for arrays 5`] = `"value is not an Array"`;

exports[`Narrowers should for arrays 6`] = `"value is not an Array"`;

exports[`Narrowers should for arrays 7`] = `"value is not an Array"`;

exports[`Narrowers should for booleans 1`] = `"value is not of type 'boolean'"`;

exports[`Narrowers should for booleans 2`] = `"value is not of type 'boolean'"`;

exports[`Narrowers should for booleans 3`] = `"value is not of type 'boolean'"`;

exports[`Narrowers should for booleans 4`] = `"value is not of type 'boolean'"`;

exports[`Narrowers should for booleans 5`] = `"value is not of type 'boolean'"`;

exports[`Narrowers should for booleans 6`] = `"value is not of type 'boolean'"`;

exports[`Narrowers should for booleans 7`] = `"value is not of type 'boolean'"`;

exports[`Narrowers should for custom classes 1`] = `"value is not an instance of C"`;

exports[`Narrowers should for custom classes 2`] = `"value is not an instance of B"`;

exports[`Narrowers should for custom classes 3`] = `"value is not an instance of D"`;

exports[`Narrowers should for custom classes 4`] = `"value is not an instance of A"`;

exports[`Narrowers should for custom classes 5`] = `"value is not an instance of B"`;

exports[`Narrowers should for custom classes 6`] = `"value is not an instance of A"`;

exports[`Narrowers should for ints (strict) 1`] = `"value is not an integer"`;

exports[`Narrowers should for ints (strict) 2`] = `"value is not an integer"`;

exports[`Narrowers should for ints (strict) 3`] = `"value is not an integer"`;

exports[`Narrowers should for ints (strict) 4`] = `"value is not an integer"`;

exports[`Narrowers should for ints (strict) 5`] = `"value is not an integer"`;

exports[`Narrowers should for ints (strict) 6`] = `"value is not an integer"`;

exports[`Narrowers should for ints (strict) 7`] = `"value is not an integer"`;

exports[`Narrowers should for ints (strict) 8`] = `"value is not an integer"`;

exports[`Narrowers should for ints (strict) 9`] = `"value is not an integer"`;

exports[`Narrowers should for ints (strict) 10`] = `"value is not an integer"`;

exports[`Narrowers should for ints (strict) 11`] = `"value is not an integer"`;

exports[`Narrowers should for ints (strict) 12`] = `"value is not an integer"`;

exports[`Narrowers should for ints (strict) 13`] = `"value is not an integer"`;

exports[`Narrowers should for ints (strict) 14`] = `"value is not an integer"`;

exports[`Narrowers should for numbers (loose) 1`] = `"value is not of type 'number'"`;

exports[`Narrowers should for numbers (loose) 2`] = `"value is not of type 'number'"`;

exports[`Narrowers should for numbers (loose) 3`] = `"value is not of type 'number'"`;

exports[`Narrowers should for numbers (loose) 4`] = `"value is not of type 'number'"`;

exports[`Narrowers should for numbers (loose) 5`] = `"value is not of type 'number'"`;

exports[`Narrowers should for numbers (loose) 6`] = `"value is not of type 'number'"`;

exports[`Narrowers should for numbers (loose) 7`] = `"value is not of type 'number'"`;

exports[`Narrowers should for numbers (loose) 8`] = `"value is not of type 'number'"`;

exports[`Narrowers should for numbers (strict) 1`] = `"number is NaN"`;

exports[`Narrowers should for numbers (strict) 2`] = `"number is NaN"`;

exports[`Narrowers should for numbers (strict) 3`] = `"value is not of type 'number'"`;

exports[`Narrowers should for numbers (strict) 4`] = `"value is not of type 'number'"`;

exports[`Narrowers should for numbers (strict) 5`] = `"value is not of type 'number'"`;

exports[`Narrowers should for numbers (strict) 6`] = `"value is not of type 'number'"`;

exports[`Narrowers should for numbers (strict) 7`] = `"value is not of type 'number'"`;

exports[`Narrowers should for numbers (strict) 8`] = `"value is not of type 'number'"`;

exports[`Narrowers should for numbers (strict) 9`] = `"value is not of type 'number'"`;

exports[`Narrowers should for numbers (strict) 10`] = `"value is not of type 'number'"`;

exports[`Narrowers should for strings 1`] = `"value is not of type 'string'"`;

exports[`Narrowers should for strings 2`] = `"value is not of type 'string'"`;

exports[`Narrowers should for strings 3`] = `"value is not of type 'string'"`;

exports[`Narrowers should for strings 4`] = `"value is not of type 'string'"`;

exports[`Narrowers should for strings 5`] = `"value is not of type 'string'"`;

exports[`Narrowers should for strings 6`] = `"value is not of type 'string'"`;
